src/Sparrow.Server/Allocators.PointerTypes.cs

```diff
@@ -0,0 +1,527 @@
+            return new Span<byte>(Address, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span<byte> AsSpan(int start, int length)
+        {
+            if (start + length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+
+            return new Span<byte>((byte*)Address + start, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<byte> AsReadOnlySpan()
+        {
+            return new ReadOnlySpan<byte>(Address, (int)Size);
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<byte> AsReadOnlySpan(int length)
+        {
+            if (length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+
+            return new ReadOnlySpan<byte>(Address, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<byte> AsReadOnlySpan(int start, int length)
+        {
+            if (start + length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+
+            return new ReadOnlySpan<byte>((byte*)Address + start, length);
+        }
+
+        public string Describe()
+        {
+            if (this.IsValid)
+                return $"{{{(long)this.Address:X32}|{this.Size}|{this.BlockSize}}}";
+
+            return "{null}";
+        }
+
+        public override string ToString()
+        {
+            return Describe();
+        }
+
+        public static implicit operator BlockPointer(Pointer ptr)
+        {
+            return new BlockPointer(ptr.Address, ptr.Size, ptr.Size);
+        }
+    }
+
+    /// <summary>
+    /// The BlockPointer is a type of pointer that carries with it 2 different sizes at the same time. The pointer value
+    /// with the proper memory size allocated to it AND the actual requested size when asked. A BlockPointer created from
+    /// a straight Pointer will have the same size for the allocation size and the actual size. 
+    /// </summary>
+    [StructLayout(LayoutKind.Sequential)]
+#if !VALIDATE
+    // We do this to have the ability to perform extensive validation for pointers.
+    readonly
+#endif
+    public unsafe struct BlockPointer<T> : Meta.IDescribe, IPointerType<T> where T : struct
+    {
+        public readonly void* Address;
+
+        private readonly int _size;
+
+        public readonly int BlockSize;
+
+#if VALIDATE
+        public uint Generation;        
+#endif
+
+        [MethodImpl(MethodImplOptions.NoOptimization)]
+        static BlockPointer()
+        {
+            // If T is not a blittable type, fail on type loading unrecoverably.
+            // We need to ensure the JIT wont optimize away the call to Marshal.SizeOf<T> for correctness either now or in the future. 
```

Comment:
Can probably do that by setting a static member value here, no?

Reply:
No need that is why we use the `[MethodImpl(MethodImplOptions.NoOptimization)]` tag